### PR TITLE
[reparentutil / wrangler] Extract PlannedReparentShard logic from wrangler to PlannedReparenter struct

### PR DESCRIPTION
**Pull request by ajm188**
## Description

This PR extracts the PRS logic from `./go/vt/wrangler` to a dedicated `PlannedReparenter` struct in `./go/vt/vtctl/reparentutil`, to be shared between the legacy and new vtctl APIs in my next PR, similar to what I did for ERS in #7464.

### Some notes on commits

All the commits that begin with "wip" are just adding tests, and it's how I committed changes as I went, so it may be useful to step through that way, but before merging I'd like to squash all of those — and just those ones — into a single, "Add all the tests" commit.

## Related Issue(s)

- #7058 
- #7428 

## Checklist
- [ ] Should this PR be backported? **No**
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

## Impacted Areas in Vitess
Components that this PR will affect:

- [ ]  Query Serving
- [ ]  VReplication
- [x]  Cluster Management
- [ ]  Build/CI
- [ ]  VTAdmin

### CODE REVIEW

**Comment by doeg** [comment]
I can review this from a Go perspective but, as with https://github.com/vitessio/vitess/pull/7464, I defer to experts on the Vitess-y parts. @deepthi, @setassociative, and/or @rohit-nayak-ps would you be able to give this one a look too?

**Comment by ajm188** [comment]
Test failure is coming from a PRS endtoend test, but looks like a port reservation issue. Going to kick it again and also try to repro locally 

```
I0217 02:51:04.937384    1595 vtctlclient_process.go:149] Executing vtctlclient with command: vtctlclient -server localhost:23806 PlannedReparentShard -keyspace_shard ks/0 -new_master zone1-0000000102
    reparent_test.go:348: 
        	Error Trace:	reparent_test.go:348
        	Error:      	"E0217 02:51:12.928862    4210 main.go:67] E0217 02:51:12.928491 planned_reparenter.go:622] some replicas failed to reparent; retry PlannedReparentShard with the same new primary alias ([zone1-0000000102 tablet zone1-0000000103 failed SetMaster(zone1-0000000102): rpc error: code = Unknown desc = TabletManager.SetMaster on zone1-0000000103 error: net.Dial(/home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock) to local server failed: dial unix /home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock: connect: no such file or directory (errno 2002) (sqlstate HY000): net.Dial(/home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock) to local server failed: dial unix /home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock: connect: no such file or directory (errno 2002) (sqlstate HY000): rpc error: code = Unknown desc = TabletManager.SetMaster on zone1-0000000103 error: net.Dial(/home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock) to local server failed: dial unix /home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock: connect: no such file or directory (errno 2002) (sqlstate HY000): net.Dial(/home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock) to local server failed: dial unix /home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock: connect: no such file or directory (errno 2002) (sqlstate HY000)]) to retry failed replicas: %!v(MISSING)
        	            	PlannedReparentShard Error: rpc error: code = Unknown desc = some replicas failed to reparent; retry PlannedReparentShard with the same new primary alias (zone1-0000000102) to retry failed replicas: tablet zone1-0000000103 failed SetMaster(zone1-0000000102): rpc error: code = Unknown desc = TabletManager.SetMaster on zone1-0000000103 error: net.Dial(/home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock) to local server failed: dial unix /home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock: connect: no such file or directory (errno 2002) (sqlstate HY000): net.Dial(/home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock) to local server failed: dial unix /home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock: connect: no such file or directory (errno 2002) (sqlstate HY000): rpc error: code = Unknown desc = TabletManager.SetMaster on zone1-0000000103 error: net.Dial(/home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock) to local server failed: dial unix /home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock: connect: no such file or directory (errno 2002) (sqlstate HY000): net.Dial(/home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock) to local server failed: dial unix /home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock: connect: no such file or directory (errno 2002) (sqlstate HY000)
        	            	E0217 02:51:12.929834    4210 main.go:72] remote error: rpc error: code = Unknown desc = some replicas failed to reparent; retry PlannedReparentShard with the same new primary alias (zone1-0000000102) to retry failed replicas: tablet zone1-0000000103 failed SetMaster(zone1-0000000102): rpc error: code = Unknown desc = TabletManager.SetMaster on zone1-0000000103 error: net.Dial(/home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock) to local server failed: dial unix /home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock: connect: no such file or directory (errno 2002) (sqlstate HY000): net.Dial(/home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock) to local server failed: dial unix /home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock: connect: no such file or directory (errno 2002) (sqlstate HY000): rpc error: code = Unknown desc = TabletManager.SetMaster on zone1-0000000103 error: net.Dial(/home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock) to local server failed: dial unix /home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock: connect: no such file or directory (errno 2002) (sqlstate HY000): net.Dial(/home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock) to local server failed: dial unix /home/runner/work/vitess/vitess/vtdataroot/vt_731717628/vtroot_23801/vt_0000000103/mysql.sock: connect: no such file or directory (errno 2002) (sqlstate HY000)
        	            	" does not contain "tablet zone1-0000000103 SetMaster failed"
        	Test:       	TestReparentWithDownReplica
```

**Review by doeg — APPROVED** [review]
This change seems straightforward and looks good to me! I do suggest waiting for another +1 from one of the other reviewers, though. :) 

Truly a heroic amount of faking + testing. 🏆 Seriously good work.

File: go/vt/vtctl/reparentutil/planned_reparenter.go
Comment (deepthi) [inline go/vt/vtctl/reparentutil/planned_reparenter.go:128]: typo: preflightChecks

**Review by deepthi — APPROVED** [review]
Very nice refactor. This is so much more readable than what we had before. :100:

File: go/vt/vtctl/reparentutil/planned_reparenter.go
Comment (ajm188) [reply go/vt/vtctl/reparentutil/planned_reparenter.go]: "prelight" 😂 

nice catch, should be fixed, and i've cleaned up the commit history as well now

**Pull request merged**
